### PR TITLE
feat: Gateway Mnemonic and Common Database

### PR DESCRIPTION
**Pull request by m1sterc001guy**
Fixes: https://github.com/fedimint/fedimint/issues/4446

Newly joined federations use clients that are created from a common mnemonic. This mnemonic is generated when the gateway boots. "Legacy" federations are still supported and can be upgraded by leaving, deleting the database, then re-joining. A test for this has been added and an automated endpoint for doing this can be added in a follow up.

This PR also cleans up the old standing issue of having separate databases per joined federation inside the gateway.

Note: This only provides backup-ability of the ecash of each joined federation. The recover API still needs to be added, as well as recovering lightning contracts as described  https://github.com/fedimint/fedimint/issues/5839

### CODE REVIEW

**Comment by Kodylow** [comment]
it's so beautiful 😍

**Comment by m1sterc001guy** [comment]
Blocked on https://github.com/fedimint/fedimint/pull/5871

File: devimint/src/devfed.rs
Comment (m1sterc001guy) [inline devimint/src/devfed.rs:383]: For reviewers: these are not used anywhere meaningful and they trigger a polling loop by contacting each gateway. Just not necessary.

**Comment by m1sterc001guy** [comment]
Had a quick chat with @bradleystachurski and he explained what I was doing wrong with the new mnemonic test. Now the mnemonic test runs as part of the upgrade tests. After the next release we can probably remove it, as we will have sufficient confidence that this upgrade path works.

Upgrade tests pass: https://github.com/fedimint/fedimint/actions/runs/10479242969/job/29024449941?pr=5852

File: gateway/ln-gateway/src/lib.rs
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/lib.rs:1473]: Currently the only way to recover from a mnemonic is to delete the database or recover using a new gateway with a different data directory. Is that ok from a user experience perspective? Thoughts?

File: gateway/ln-gateway/src/client.rs
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/client.rs:115]: Looking for some feedback on this too. I used the `mint_channel_id` here because it was smaller than `federation_id` and it is also unique, but ideally I'd like to remove this when we remove LNv1. Maybe `FederationIdPrefix` is better?

Comment (Kodylow) [reply gateway/ln-gateway/src/client.rs:115]: federationIdPrefix is definitely better

File: gateway/ln-gateway/src/lib.rs
Comment (Kodylow) [reply gateway/ln-gateway/src/lib.rs:1473]: This seems fine for now, recovery in separate PR

File: gateway/integration_tests/src/main.rs
Comment (Kodylow) [inline gateway/integration_tests/src/main.rs:46]: you said this is gonna be temporary until we phase out the legacy gateway multi dir structure? can we add a todo tagging it and we'll add to the tracking issue of things to phase out?

File: gateway/integration_tests/src/main.rs
Comment (Kodylow) [inline gateway/integration_tests/src/main.rs:73]: same here, tag TODO for removing

**Review by Kodylow — COMMENTED** [review]
lgtm can you add instructions for local testing? also think we'll need to surface the mnemonic behind a protected path on the rpc server so I can reveal in the gateway for when someone creates it locally and they want to back it up

**Comment by m1sterc001guy** [comment]
> can you add instructions for local testing?

Sure. Running the upgrade tests should be sufficient to verify it works (`./scripts/tests/upgrade-test.sh v0.4.0`). Version here doesn't matter for the mnemonic test. You can also test it in mprocs (`gateway-lnd seed`) to see the seed phrase for the various gateways.

> also think we'll need to surface the mnemonic behind a protected path on the rpc server so I can reveal in the gateway for when someone creates it locally and they want to back it up

The `/mnemonic` path is already a protected path. You can verify in mprocs.

File: gateway/integration_tests/src/main.rs
Comment (m1sterc001guy) [reply gateway/integration_tests/src/main.rs:46]: The test is really only needed until the next release, at which point we've essentially verified that this upgrade path works. We don't need to re-run a test with old code over and over again. Yes, I will add a comment todo tag

File: modules/fedimint-ln-common/src/gateway_endpoint_constants.rs
Comment (Kodylow) [inline modules/fedimint-ln-common/src/gateway_endpoint_constants.rs:9]: oh sorry didn't see you'd added this already

File: gateway/integration_tests/src/main.rs
Comment (m1sterc001guy) [reply gateway/integration_tests/src/main.rs:46]: Added TODO

File: gateway/ln-gateway/src/client.rs
Comment (m1sterc001guy) [reply gateway/ln-gateway/src/client.rs]: Changed to use `FederationIdPrefix`

File: devimint/src/gatewayd.rs
Comment (tvolk131) [inline devimint/src/gatewayd.rs:239]: Can you explain why this needs to be wrapped in a `retry()` call?

File: gateway/ln-gateway/src/rpc/rpc_server.rs
Comment (tvolk131) [inline gateway/ln-gateway/src/rpc/rpc_server.rs:191]: Phew, good thing this is authenticated 😮‍💨

File: gateway/ln-gateway/src/rpc/mod.rs
Comment (tvolk131) [inline gateway/ln-gateway/src/rpc/mod.rs:236]: Can you add a comment explaining what `legacy_federations` does and why it exists?

File: modules/fedimint-ln-common/src/gateway_endpoint_constants.rs
Comment (tvolk131) [inline modules/fedimint-ln-common/src/gateway_endpoint_constants.rs:9]: Nit: Is the `GATEWAY_` prefix necessary here? The file is prefixed with `gateway_` already.

File: fedimint-core/src/config.rs
Comment (dpc) [inline fedimint-core/src/config.rs:374]: Idiomatically, if it allocates, it should be `to_`. `as_` is only for near-zero-cost conversions.

**Review by dpc — COMMENTED** [review]
I looked through it very lightly and LGTM.

File: gateway/ln-gateway/src/client.rs
Comment (joschisan) [inline gateway/ln-gateway/src/client.rs:145]: Why not filter an iterator here?

File: devimint/src/gatewayd.rs
Comment (m1sterc001guy) [reply devimint/src/gatewayd.rs:239]: In the test, this function is used right after restarting/upgrading the gateway. To avoid flakiness (i.e requesting the mnemonic when the webserver isnt up yet), it is easiest to just wrap in a retry.

File: gateway/ln-gateway/src/client.rs
Comment (m1sterc001guy) [reply gateway/ln-gateway/src/client.rs]: Done

File: fedimint-core/src/config.rs
Comment (m1sterc001guy) [reply fedimint-core/src/config.rs]: Done

File: modules/fedimint-ln-common/src/gateway_endpoint_constants.rs
Comment (m1sterc001guy) [reply modules/fedimint-ln-common/src/gateway_endpoint_constants.rs]: Done

File: gateway/ln-gateway/src/rpc/mod.rs
Comment (m1sterc001guy) [reply gateway/ln-gateway/src/rpc/mod.rs:247]: Done

**Review by tvolk131 — DISMISSED** [review]
LGTM! Should get another approval from someone with more familiarity before submitting

**Comment by m1sterc001guy** [comment]
Ah shoot, newly added `gatewayd-mnemonic` test is failing. I will fix before trying to merge.

**Comment by m1sterc001guy** [comment]
> Ah shoot, newly added gatewayd-mnemonic test is failing. I will fix before trying to merge.

Likely related to https://github.com/fedimint/fedimint/issues/5948

**Comment by bradleystachurski** [comment]
> > Ah shoot, newly added gatewayd-mnemonic test is failing. I will fix before trying to merge.
> 
> Likely related to #5948

I still don't know the root cause, but reverting this commit fixed the upgrade tests locally. You can try a revert on your end to see if it's the same issue.

https://github.com/fedimint/fedimint/commit/11d61ad2046468acaf326dd63430e7b64906a62c

File: fedimint-core/src/util/mod.rs
Comment (elsirion) [inline fedimint-core/src/util/mod.rs:444]: For `error` the `Debug` printing makes sense, less so for `attempts`.

File: gateway/ln-gateway/src/client.rs
Comment (elsirion) [inline gateway/ln-gateway/src/client.rs:54]: Do we still need this branch in `client_plainrootsecret`? I'd imagine all new clients use the global secret, which is a BIP39 one and not plain?

File: gateway/ln-gateway/src/client.rs
Comment (elsirion) [inline gateway/ln-gateway/src/client.rs:115]: We directly return the root secret here, but should be deriving a per-federation secret as described in the [secret derivation docs](https://github.com/fedimint/fedimint/blob/master/docs/secret_derivation.md).

The current version isn't *horribly* broken because out of abundance of caution clients internally derive a key based on the federation id, but to be compatible with other wallet tools using the standard derivation path would be nice.

File: gateway/ln-gateway/src/lib.rs
Comment (elsirion) [reply gateway/ln-gateway/src/lib.rs:1566]: In a follow-up PR it would be nice to make setting the mnemonic an API call.

File: gateway/ln-gateway/src/client.rs
Comment (m1sterc001guy) [reply gateway/ln-gateway/src/client.rs:115]: Hmm interesting, in early discussions with @dpc I thought the client was already deriving a per-federation secret using the federation id, and that adding it here would actually be doing the derivation twice unnecessarily. I actually originally had it like that, but ended up removing the derivation after our discussion. 

Is it recommended to derive a per-federation secret here?

File: gateway/ln-gateway/src/client.rs
Comment (m1sterc001guy) [reply gateway/ln-gateway/src/client.rs]: Correct, not necessary, good catch. Removed.

File: fedimint-core/src/util/mod.rs
Comment (m1sterc001guy) [reply fedimint-core/src/util/mod.rs]: Changed attempts back

File: gateway/ln-gateway/src/client.rs
Comment (elsirion) [reply gateway/ln-gateway/src/client.rs:102]: We are doing the derivation twice if following the standard, so if an integrator doesn't follow the standard then they don't shoot themselves (or rather their users) in the foot. 

The reasoning was: 
1. It doesn't degrade security 
2. The time it takes is negligible
3. Integrators won't know about the second client-internal derivation, so it won't feel weird to them

Comment (m1sterc001guy) [reply gateway/ln-gateway/src/client.rs]: Fixed

File: gateway/integration_tests/src/main.rs
Comment (bradleystachurski) [inline gateway/integration_tests/src/main.rs:77]: [nit] It might be useful to duplicate this TODO in `upgrade-test.sh`

File: gateway/ln-gateway/src/client.rs
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/client.rs:100]: Use `mint_channel_id` here, also add check to make sure we don't re-use a database for multiple clients

Comment (m1sterc001guy) [reply gateway/ln-gateway/src/client.rs]: Done. Also added a check that verifies the stored `ClientConfig`'s federation ID matches the federation ID we are creating. This should catch any accidental re-use of databases.

**Review by elsirion — APPROVED** [review]

**Comment by fedimint-backports** [comment]
Backport failed for `releases/v0.4`, because it was unable to cherry-pick the commit(s).

Please cherry-pick the changes locally and resolve any conflicts.
```bash
git fetch origin releases/v0.4
git worktree add -d .worktree/backport-5852-to-releases/v0.4 origin/releases/v0.4
cd .worktree/backport-5852-to-releases/v0.4
git switch --create backport-5852-to-releases/v0.4
git cherry-pick -x 13df2f0682cdf74073561085465d17734d1b5c69 532c66e8a5c74fdf6d7eb3eea967cb82024d6bfa 81f769f5761b2b2f78565078b673d11702e236d0 128d475986762c5c547f3f01fba5b182ef6fd6bd a0ddae713dc493171ad3b27705b224268fb6aa33
```

**Pull request merged**
